# Costa Rican Household Poverty Level Prediction

Costa Rican Household Poverty Level Prediction is a Kaggle competition which is currently open for late submissions. This competition is for a social cause to help Inter-American Development Bank to identify which households have the highest need for social welfare assistance. Many social programs have a hard time making sure the right people are given enough aid. It’s especially tricky when a program focuses on the poorest segment of the population. The world’s poorest typically can’t provide the necessary income and expense records to prove that they qualify. Other than Costa Rica, many other countries also face this same problem of inaccurately assessing social need. If data scientists can generate an improvement, many countries can take benefit out of it.

The dataset available in Kaggle is a set of household characteristics from a representative sample of Costa Rican Households. The dataset has observations for each member of the household but the classification is done at the household level. This is a multi-class classification problem using a supervised machine learning approach. We classify households according to their income levels (1 = extreme poverty,  2 = moderate poverty, 3 = vulnerable households, 4 = non-vulnerable households). We have identified the multiclass classification algorithms which we think apt to use this type of problems like, KNN, Logistic Regression, Naive Bayes, Random Forest etc. We will fine-tune the hyperparameters for each model and evaluate the different models using f1-score and accuracy metrics to find out the top best among them. Will do significance tests to figure out the optimal model to perform prediction.

Kaggle competition link:

https://www.kaggle.com/c/costa-rican-household-poverty-prediction (Links to an external site.)Links to an external site.
  

# Setup

First, let's make sure this notebook works well in both python 2 and 3, import a few common modules, ensure MatplotLib plots figures inline.

In [28]:
# To support both python 2 and python 3
from __future__ import division, print_function, unicode_literals

# Common imports
import numpy as np
import os

# to make this notebook's output stable across runs
np.random.seed(42)

# To plot pretty figures
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
plt.rcParams['axes.labelsize'] = 14
plt.rcParams['xtick.labelsize'] = 12
plt.rcParams['ytick.labelsize'] = 12

# Download the data and do some verifications

The data is located locally within this folder. 

In [29]:
%ls -l all/train.csv

-rwxr-xr-x 1 root root 3237288 Jul 19 03:51 all/train.csv*


In [30]:
import pandas as pd

In [31]:
full_train_data = pd.read_csv('all/train.csv')

In [32]:
train_data = pd.read_csv('all/train.csv')
test_data = pd.read_csv('all/test.csv')

In [33]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 143 entries, Id to Target
dtypes: float64(8), int64(130), object(5)
memory usage: 10.4+ MB


In [34]:
pd.options.display.max_columns = 150
train_data.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       NaN       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     0     1     1     0     0     0     0     1     1       1       1   
1     0     1     1     0     0     0     0     1     1       1       1   
2     0     0     0     0     1     1     0     1     1       1       1   
3     0     2     2     1     1     2     1     3     4       4       4   
4     0     2     2     1     1     2     1     3     4       4       4   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0        10      NaN       1            1            0          0         0   
1        12      NaN       1            0            0          0         0   
2        11      NaN       1            0            0          0         0   
3         9      1.0       4            1            0          0         0   
4        11      NaN       4            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         1          0            0           0           0            0   
2         1          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          0               1   
1          0          0            0           1          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              1   
2           0           0           0              0              1   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              0              0          1          0          0          0   
2              0              0          1          0          0          0   
3              0              0          1          0

In [35]:
train_data.corr()

v2a1    hacdor     rooms    hacapo      v14a    refrig  \
v2a1             1.000000 -0.091732  0.443461 -0.073509  0.033551  0.088970   
hacdor          -0.091732  1.000000 -0.233369  0.652594 -0.175011 -0.101965   
rooms            0.443461 -0.233369  1.000000 -0.213368  0.129183  0.130531   
hacapo          -0.073509  0.652594 -0.213368  1.000000 -0.150986 -0.124506   
v14a             0.033551 -0.175011  0.129183 -0.150986  1.000000  0.143143   
refrig           0.088970 -0.101965  0.130531 -0.124506  0.143143  1.000000   
v18q             0.278364 -0.084680  0.254256 -0.067529  0.036396  0.086002   
v18q1            0.302292 -0.049262  0.208919 -0.037414  0.011255 -0.070318   
r4h1            -0.081900  0.232508 -0.066578  0.226378 -0.054769 -0.047087   
r4h2            -0.002401  0.059313  0.267627  0.126645  0.018133 -0.022819   
r4h3            -0.059548  0.184857  0.195222  0.240056 -0.015552 -0.046860   
r4m1            -0.072791  0.268978 -0.032558  0.241452 -0.006370 -0.023502   
r4m2             0.048169  0.142458  0.241989  0.095545  0.038997  0.027832   
r4m3            -0.007806  0.264620  0.168503  0.212527  0.026510  0.008038   
r4t1            -0.103959  0.328901 -0.064789  0.306722 -0.039803 -0.046136   
r4t2             0.032631  0.134909  0.349206  0.152968  0.038295  0.001607   
r4t3            -0.043574  0.304282  0.245784  0.305857  0.007615 -0.025979   
tamhog          -0.064566  0.304884  0.240137  0.306289  0.007328 -0.026784   
tamviv          -0.051396  0.350948  0.254473  0.354389  0.010613 -0.022060   
escolari         0.288006 -0.122134  0.220284 -0.104846  0.036796  0.097733   
rez_esc         -0.086238  0.047466 -0.084857  0.081574 -0.022039 -0.116235   
hhsize          -0.064566  0.304884  0.240137  0.306289  0.007328 -0.026784   
paredblolad      0.290062 -0.136055  0.260981 -0.126550  0.058186  0.134937   
paredzocalo     -0.101152  0.074927 -0.057740  0.057683 -0.000611 -0.035913   
paredpreb       -0.174037 -0.027202 -0.151392 -0.036119  0.005202  0.022801   
pareddes        -0.037311  0.147449 -0.096855  0.164714 -0.134787 -0.087432   
paredmad        -0.134678  0.083620 -0.111619  0.083705 -0.059958 -0.162714   
paredzinc       -0.053664  0.155120 -0.070527  0.169976  0.008349 -0.053502   
paredfibras     -0.025562 -0.007622 -0.002567 -0.005961  0.002778  0.008057   
paredother      -0.015824 -0.007622 -0.021199 -0.005961  0.002778  0.008057   
pisomoscer       0.296623 -0.168472  0.266992 -0.166538  0.086897  0.169537   
pisocemento     -0.273600  0.077334 -0.203644  0.073848 -0.020410 -0.068315   
pisoother             NaN -0.006109  0.003253 -0.004778  0.002227  0.006459   
pisonatur             NaN  0.078103 -0.040900 -0.005037 -0.221964 -0.073521   
pisonotiene     -0.031581  0.232319 -0.142969  0.256174 -0.059090 -0.099486   
pisomadera      -0.085837  0.055087 -0.076674  0.055785 -0.068411 -0.140878   
techozinc       -0.001022 -0.042053  0.044452 -0.069665  0.063877  0.014877   
techoentrepiso   0.011072 -0.026698 -0.033245 -0.020881  0.009730  0.020343   
techocane       -0.029893 -0.011166  0.002973 -0.008733  0.004070  0.011805   
techootro             NaN -0.009338  0.066864 -0.007303  0.003403  0.009872   
cielorazo        0.300238 -0.177257  0.310513 -0.169545  0.080269  0.171503   
abastaguadentro  0.055717 -0.053838  0.078825 -0.056118  0.151276  0.111854   
abastaguafuera  -0.055717  0.060714 -0.066509  0.062234 -0.119326 -0.083479   
abastaguano           NaN -0.011890 -0.048460 -0.009299 -0.117470 -0.100841   
public           0.121159  0.018548  0.052554 -0.000075  0.014807  0.064289   
planpri               NaN -0.003526 -0.003486 -0.002758  0.001285 -0.025594   
noelec                NaN -0.009338 -0.054890 -0.007303 -0.089524 -0.223078   
coopele         -0.121159 -0.016256 -0.042660  0.002100  0.020980 -0.011786   
sanitario1      -0.027800  0.102032 -0.125488  0.134452 -0.532663 -0.078890   
sanitario2       0.141261 -0.018513  0.076446 -0.032506  0.006009  0.00

### Number of Null Columns.

In [36]:
pd.options.display.max_rows = 150
train_data.isnull().sum()

Id                    0
v2a1               6860
hacdor                0
rooms                 0
hacapo                0
v14a                  0
refrig                0
v18q                  0
v18q1              7342
r4h1                  0
r4h2                  0
r4h3                  0
r4m1                  0
r4m2                  0
r4m3                  0
r4t1                  0
r4t2                  0
r4t3                  0
tamhog                0
tamviv                0
escolari              0
rez_esc            7928
hhsize                0
paredblolad           0
paredzocalo           0
paredpreb             0
pareddes              0
paredmad              0
paredzinc             0
paredfibras           0
paredother            0
pisomoscer            0
pisocemento           0
pisoother             0
pisonatur             0
pisonotiene           0
pisomadera            0
techozinc             0
techoentrepiso        0
techocane             0
techootro             0
cielorazo       

 ### Columns with Null values arranged descending. 

1. rez_esc -- 7928 
2. v18q1   -- 7342
3. v2a1    -- 6860
4. meaneduc -- 5
5. SQBmeaned -- 5
    
    

In [38]:
train_data['target'] = train_data['Target']
train_data['target']=train_data['target'].apply(str)

In [39]:
train_data['target'] = train_data['target'].map({'1': 'extreme poverty' ,'2' : 'moderate poverty', 
                              '3' :'vulnerable households', '4' : 'non vulnerable households'})

In [40]:
train_data.target.value_counts()

non vulnerable households    5996
moderate poverty             1597
vulnerable households        1209
extreme poverty               755
Name: target, dtype: int64

In [44]:
pd.options.display.max_rows = 150
test_data.isnull().sum()

Id                     0
v2a1               17403
hacdor                 0
rooms                  0
hacapo                 0
v14a                   0
refrig                 0
v18q                   0
v18q1              18126
r4h1                   0
r4h2                   0
r4h3                   0
r4m1                   0
r4m2                   0
r4m3                   0
r4t1                   0
r4t2                   0
r4t3                   0
tamhog                 0
tamviv                 0
escolari               0
rez_esc                0
hhsize                 0
paredblolad            0
paredzocalo            0
paredpreb              0
pareddes               0
paredmad               0
paredzinc              0
paredfibras            0
paredother             0
pisomoscer             0
pisocemento            0
pisoother              0
pisonatur              0
pisonotiene            0
pisomadera             0
techozinc              0
techoentrepiso         0
techocane              0


In [ ]:
 ### Columns with Null values arranged descending. 

1. v18q1   -- 18126
2. v2a1    -- 17403
3. meaneduc -- 31
4. SQBmeaned -- 31

# DATA PRE_PROCESSING

## Handling Missing Data

v18q --  owns a tablet
v18q1 -- number of tablets household owns

v18q1-- Train data have around 7342 missing values. We can try to compare this missing values with the v18q to see if any of these owns a tables. If none then we can replace the value directly with 0.


In [61]:
## v18q1  --- This is number of tablet owned by a family.

heads_train = train_data.loc[train_data['parentesco1'] == 1].copy()
heads_train.groupby('v18q')['v18q1'].apply(lambda x: x.isnull().sum())

##This shows that every member have null v18q1 have 0 in v18q. So we can safely replace the 



v18q
0    2318
1       0
Name: v18q1, dtype: int64

In [60]:
heads_test = test_data.loc[test_data['parentesco1'] == 1].copy()
heads_test.groupby('v18q')['v18q1'].apply(lambda x: x.isnull().sum())


v18q
0    5726
1       0
Name: v18q1, dtype: int64

In [63]:
train_data['v18q1'] = train_data['v18q1'].fillna(0)
test_data['v18q1'] = test_data['v18q1'].fillna(0)

### rez_esc --- Years behind school. 

In [64]:
train_data['rez_esc'] = train_data['rez_esc'].fillna(0)
test_data['rez_esc'] = test_data['rez_esc'].fillna(0)

In [ ]:
### v2a1 -- 'Monthly rent payment' 
There could be known reason for missing monthly payment or missing values here. 

1. The person can have his own home.We will set the value to 0 for such cases. 
2. Other missing due to unknown reason, we will set the value to  

